src/OrleansConsulUtils/ConsulBasedMembershipTable.cs

```diff
@@ -0,0 +1,215 @@
+    /// </summary>
+    public class ConsulBasedMembershipTable : IMembershipTable, IGatewayListProvider
+    {
+        //Consul does not support the extended Membership Protocol and will always return the same table version information
+        private readonly TableVersion _tableVersion = new TableVersion(0, "0");
+
+        private TraceLogger _logger;
+        private ConsulClient _consulClient = null;
+        private String _deploymentId;
+        private String _connectionString;
+
+        private TimeSpan _maxStaleness;
+
+        public TimeSpan MaxStaleness
+        {
+            get { return _maxStaleness; }
+        }
+
+        public Boolean IsUpdatable
+        {
+            get { return true; }
+        }
+
+        public Task InitializeGatewayListProvider(ClientConfiguration config, TraceLogger traceLogger)
+        {
+            _logger = traceLogger;
+            _deploymentId = config.DeploymentId;
+            _connectionString = config.DataConnectionString;
+            _maxStaleness = config.GatewayListRefreshPeriod;
+
+            _consulClient =
+                new ConsulClient(
+                    new ConsulClientConfiguration
+                    {
+                        Address = new Uri(config.DataConnectionString)
+                    });
+
+            return TaskDone.Done;
+        }
+
+        /// <summary>
+        /// Initializes the Consul based membership table.
+        /// </summary>
+        /// <param name="config">The configuration for this instance.</param>
+        /// <param name="tryInitPath"></param>
+        /// <param name="traceLogger">The logger to be used by this instance</param>
+        /// <returns></returns>
+        /// <remarks>
+        /// Consul Membership Provider does not support the extended Membership Protocol,
+        /// therefore there is no MembershipTable to Initialise
+        /// </remarks>
+        public Task InitializeMembershipTable(GlobalConfiguration config, Boolean tryInitPath, TraceLogger traceLogger)
+        {
+            _logger = traceLogger;
+            _deploymentId = config.DeploymentId;
+            _connectionString = config.DataConnectionString;
+
+            _consulClient =
+                new ConsulClient(
+                    new ConsulClientConfiguration
+                    {
+                        Address = new Uri(config.DataConnectionString)
+                    });
+
+            return TaskDone.Done;
+        }
+
+        public async Task<MembershipTableData> ReadRow(SiloAddress siloAddress)
+        {
+            var siloRegistration = await GetConsulSiloRegistration(siloAddress);
+
+            return AssembleMembershipTableData(siloRegistration);
+        }
+
+        public async Task<MembershipTableData> ReadAll()
+        {
+            var deploymentKVAddresses = await _consulClient.KV.List(ConsulSiloRegistration.ParseDeploymentKVPrefix(_deploymentId));
```

Comment:
so those `_consulClient.KV.List` and `_consulClient.KV.Get` are async, right? Just making sure, since the usual pattern is to call them `GetAsync`, `ListAsync`.